x/ibc/03-connection/types/version.go

```diff
@@ -1,96 +1,171 @@
-	case n == 0 || m == 0:
-		return ""
-	case n == m:
-		for i := n - 1; i >= 0; i-- {
-			counterpartyVerLookup[counterpartyVersions[i]] = true
-			compatibleVerLookup[compatibleVersions[i]] = true
+	// check if features are empty
+	if len(featureSet) == 0 {
+		return identifier, []string{}, nil
+	}
 
-			// check if we've seen any of the versions
-			if _, ok := compatibleVerLookup[counterpartyVersions[i]]; ok {
-				supportedVersions = append(supportedVersions, counterpartyVersions[i])
-			}
+	features := strings.Split(featureSet, ",")
 
-			if _, ok := counterpartyVerLookup[compatibleVersions[i]]; ok {
-				// TODO: check if the version is already in the array
-				supportedVersions = append(supportedVersions, compatibleVersions[i])
-			}
-		}
-	case n > m:
-		for i := n - 1; i >= m; i-- {
-			counterpartyVerLookup[compatibleVersions[i]] = true
+	return identifier, features, nil
+}
+
+// FindSupportedVersion returns the version with a matching version identifier
+// if it exists. The returned boolean is true if the version is found and
+// false otherwise.
+func FindSupportedVersion(version string, supportedVersions []string) (string, bool) {
+	identifier, _, err := UnpackVersion(version)
+	if err != nil {
+		return "", false
+	}
+
+	for _, supportedVersion := range supportedVersions {
+		supportedIdentifier, _, err := UnpackVersion(supportedVersion)
+		if err != nil {
+			continue
 		}
 
-		for i := m - 1; i >= 0; i-- {
-			counterpartyVerLookup[counterpartyVersions[i]] = true
-			compatibleVerLookup[compatibleVersions[i]] = true
+		if identifier == supportedIdentifier {
+			return supportedVersion, true
+		}
+	}
+	return "", false
+}
 
-			// check if we've seen any of the versions
-			if _, ok := compatibleVerLookup[counterpartyVersions[i]]; ok {
-				supportedVersions = append(supportedVersions, counterpartyVersions[i])
+// PickVersion iterates over the descending ordered set of compatible IBC
+// versions and selects the first version with a version identifier that is
+// supported by the counterparty. The returned version contains a feature
+// set with the intersection of the features supported by the source and
+// counterparty chains. This function is called in the ConnOpenTry handshake
+// procedure.
+func PickVersion(counterpartyVersions []string) (string, error) {
+	versions := GetCompatibleVersions()
+
+	for _, ver := range versions {
+		// check if the source version is supported by the counterparty
+		if counterpartyVer, found := FindSupportedVersion(ver, counterpartyVersions); found {
+			sourceIdentifier, sourceFeatures, err := UnpackVersion(ver)
+			if err != nil {
+				return "", err
 			}
 
-			if _, ok := counterpartyVerLookup[compatibleVersions[i]]; ok {
-				supportedVersions = append(supportedVersions, compatibleVersions[i])
+			_, counterpartyFeatures, err := UnpackVersion(counterpartyVer)
+			if err != nil {
+				return "", err
 			}
+
+			featureSet := GetFeatureSetIntersection(sourceFeatures, counterpartyFeatures)
```

Comment:
What happens if this intersection is null?

Looking at the code, it looks like the handshake will fail because the featureset is not valid.

However, there may be a valid feature set on a less preferred version.

Ex:

connectionA versions: {"1.1.0-ORDERED", "1.0.0-ORDERED,UNORDERED"}
connectionB versions: {"1.1.0-UNORDERED", "1.0.0-ORDERED,UNORDERED"}

The code here would negotiate to the version string "1.1.0-", and then fail because of the null featureset in ACK. It should instead negotiate to "1.0.0-ORDERED,UNORDERED"

Instead if the feature set is null, we should continue searching the lower versions to see if theres a valid interesection there.

```suggestion
			featureSet := GetFeatureSetIntersection(sourceFeatures, counterpartyFeatures)
			if featureSet == nil {
			    continue
			}
```

Reply:
Please include a test of the above case as well

Reply:
I think for now we should allow zero-intersection?

Reply:
> I think for now we should allow zero-intersection?

wouldn't that be as if the versions were totally incompatible?

Reply:
I considered this in design. If a party would like to prioritize maximum size feature set versions, they should order their versioning to reflect so (oldest version to latest, in all likelyhood). I think trying to add extra logic continuing to look for a non-null feature set is overkill for 1.0 and could be added (if ever needed). 

Furthermore, given this version negotiation system, an assumption that can be made is the core version corresponds to an implied set of core features that should be implemented by both parties. Therefore, an agreed upon version identifier with a nil feature set should still be usable.

I'm sure there would be many cases where I'd like to use the latest version with a nil feature set over using any older version (perhaps for performance reasons)

Reply:
test case for this exists:
`{"valid identifier match but empty feature set", []string{"(1,[DAG,ORDERED-ZK,UNORDERED-zk])"}, "(1,[])", true},`